### PR TITLE
Force ReJIT before 1st execution of a method

### LINKED ISSUE #1242: ReJIT race: 1st time execution may not be instrumented - detected via StrongNamedTests
[Edit 01: updated the description to reflect the actual underlying issue]

The bytecode instrumentations are added via [requests to ReJIT](https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo4-requestrejit-method) the targeted methods. Those requests are made when a module finishes loading at the callback [MethodLoadedFinished](https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback-moduleloadfinished-method). The actual ReJIT request happens in a dedicated thread (per docs recommendation) and until such request is completed there is a time window in which the method can be JIT compiled and executed without the expected bytecode instrumentation. The instrumentation is going to happen only on the first method invocation after the ReJIT request is completed.

4 instances of test failures showing this issue:

- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/3069538719/jobs/4958307009#step:4:2590
- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/3069538719/jobs/4958306796#step:4:2682
- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/3069538719/jobs/4958306876#step:4:3392
- https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/3062693118/jobs/4943951142#step:4:3392

### PR DESCRIPTION
**Pull request by pjanotti**
## Why

Enqueueing requests to ReJIT can cause the first executions of a method targeted for bytecode instrumentation to not be instrumented.

Fixes #1242

## What

Make the request to ReJIT a method targeted for bytecode instrumentation to be made on the same thread finishing the module load targeted for instrumentation. This ensures that any execution of the targeted methods is going to use the ReJIT-ed version of the method. 

## Tests

Manual runs on mac and Windows without hitting the error - these errors have low frequency on more powerful machines but runs in ci and verify-test workflows were also good.

## Checklist

- [X] `CHANGELOG.md` is updated.
~~- [ ] Documentation is updated.~~
~~- [ ] New features are covered by tests.~~

### CODE REVIEW

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.cpp [hunk 1]
```diff
@@ -370,35 +288,41 @@ void RejitHandler::AddNGenModule(ModuleID moduleId)
...
-}
+    HRESULT hr;
+    auto profilerInfo = m_profilerInfo;
+    auto profilerInfo10 = m_profilerInfo10;
 
-void RejitHandler::Shutdown()
-{
-    m_rejit_queue->push(RejitItem::CreateEndRejitThread());
-    if (m_rejit_queue_thread->joinable())
+    if (profilerInfo10 != nullptr)
```
Comment (Kielek) [inline src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.cpp:305]: What is the reason to have this condition here?
`m_profilerInfo` is `ICorProfilerInfo4` which is interface containing `RequestReJIT` fuction https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo4-requestrejit-method.

Do you want to use here `RequestReJITWithInliners ` https://learn.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo10-requestrejitwithinliners-method or you have some other reason?

**Comment by RassK** [comment]
Any ideas what was the original idea about queuing rejit requests? Any possible skeletons when changing from async to sync?
(Just about to gain some knowledge on this)

**Comment by pjanotti** [comment]
@RassK 
> Any ideas what was the original idea about queuing rejit requests? Any possible skeletons when changing from async to sync?

The docs state that requesting a ReJIT can cause a GC plus if requesting for everything (or almost everything) in a module it could be an extensive list. In sync mode that can delay finishing the ModuleLoadFinished thread and no code from that module will be executed until all requests are made. In our case, we typically instrument a small number of methods of a module, and more importantly, we need to be consistent and not miss a single execution of the method, i.e.: I don't think we can allow that to happen in name of a perhaps slightly quicker time from module load to first method execution.

**Review by nrcventura — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.cpp (hunk 1 above)
Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.cpp:305]: Good point @Kielek, ICorProfilerInfo4 is .NET Framework 4.5 so we don't need to worry about that version anymore, we can consider that we have ICorProfilerInfo6 and above. Cleaning it up...

Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation.Native/rejit_handler.cpp:305]: We can remove this multiple profiler interface members but it is better to do in a separate PR. I will do it as soon as this one is merged.

**Pull request merged**